packages/ketcher-react/src/script/editor/utils/canvasExtension.ts

```diff
@@ -0,0 +1,209 @@
+  }
+  const { leftX, topY, rightX, bottomY } = selectedAreaCoordinates;
+  const canvasXOffset = canvasBoundingRect.x - clientAreaBoundingRect.x;
+  const canvasYOffset = canvasBoundingRect.y - clientAreaBoundingRect.y;
+  const isCloseToTopEdgeOfScreen = topY + canvasYOffset <= edgeOffset;
+  const isCloseToBottomEdgeOfScreen =
+    clientArea.clientHeight - (bottomY + canvasYOffset) <= edgeOffset;
+  const isCloseToLeftEdgeOfScreen = leftX + canvasXOffset <= edgeOffset;
+  const isCloseToRightEdgeOfScreen =
+    clientArea.clientWidth - (rightX + canvasXOffset) <= edgeOffset;
+  return {
+    isCloseToLeftEdgeOfScreen,
+    isCloseToTopEdgeOfScreen,
+    isCloseToRightEdgeOfScreen,
+    isCloseToBottomEdgeOfScreen,
+  };
+}
+
+export function calculateCanvasExtension(
+  clientArea,
+  currentCanvasSize,
+  extensionVector,
+) {
+  const newHorizontalScrollPosition = clientArea.scrollLeft + extensionVector.x;
+  const newVerticalScrollPosition = clientArea.scrollTop + extensionVector.y;
+  let horizontalExtension = 0;
+  let verticalExtension = 0;
+  if (newHorizontalScrollPosition > currentCanvasSize.x) {
+    horizontalExtension = newHorizontalScrollPosition - currentCanvasSize.x;
+  }
+  if (newHorizontalScrollPosition < 0) {
+    horizontalExtension = Math.abs(newHorizontalScrollPosition);
+  }
+  if (newVerticalScrollPosition > currentCanvasSize.y) {
+    verticalExtension = newVerticalScrollPosition - currentCanvasSize.y;
+  }
+  if (newVerticalScrollPosition < 0) {
+    verticalExtension = Math.abs(newVerticalScrollPosition);
+  }
+  return new Vec2(horizontalExtension, verticalExtension, 0);
+}
+
+export function shiftByVector(vector: Vec2, editor: Editor) {
+  const render = editor.render;
+  const clientArea = render.clientArea;
+  const extensionVector = calculateCanvasExtension(
+    clientArea,
+    render.sz.scaled(render.options.zoom),
+    vector,
+  ).scaled(1 / render.options.zoom);
+
+  if (extensionVector.x > 0 || extensionVector.y > 0) {
+    // When canvas extends previous (0, 0) coordinates may become (100, 100)
+    lastX += extensionVector.x;
+    lastY += extensionVector.y;
+  }
+  render.update(false);
+  scrollToEdgeOfScreen(vector, render);
+}
+
+export function scrollByVector(vector: Vec2, render) {
+  requestAnimationFrame(() => {
+    const clientArea = render.clientArea;
+    clientArea.scrollLeft += vector.x * scrollMultiplier;
+    clientArea.scrollTop += vector.y * scrollMultiplier;
+  });
+}
+
+export function scrollToEdgeOfScreen(vector: Vec2, render) {
+  requestAnimationFrame(() => {
+    const clientArea = render.clientArea;
+    if (vector.x < 0) {
+      clientArea.scrollLeft = 0;
+    }
+    if (vector.x > 0) {
+      clientArea.scrollLeft = clientArea.scrollWidth;
+    }
+    if (vector.y < 0) {
+      clientArea.scrollTop = 0;
+    }
+    if (vector.y > 0) {
+      clientArea.scrollTop = clientArea.scrollHeight;
+    }
+  });
+}
+
+export function scrollByDirection(editor: Editor, direction) {
```

Comment:
When structure is near the edge of screen (not canvas) and is moved out of visible area, clientArea is scrolled to follow the structure. You can influence on scroll speed using scrollMultiplier